### PR TITLE
[code-completion] Add type context for single-expression function bodies

### PR DESCRIPTION
**Pull request by benlangmuir**
Extend the support for single-expression closures to handle single-expression functions of all kinds. This allows, e.g.

    func foo() -> MyEnum { .<here> }

to complete members of `MyEnum`.

rdar://problem/48938531

---

Note: this conceptually depends on https://github.com/apple/swift/pull/23251, which implements the language feature

### CODE REVIEW

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@nate-chandler FYI this PR adds code-completion support for your "allow return to be omitted" change.  The implementation doesn't require your changes so far, so I've left it a separate PR to make it easier to review.

File: lib/IDE/ExprContextAnalysis.cpp [hunk 1]
```diff
@@ -695,6 +695,12 @@ class ExprContextAnalyzer {
       break;
     }
     default:
+      if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
+        assert(isSingleExpressionBodyForCodeCompletion(AFD->getBody()));
+        singleExpressionBody = true;
+        recordPossibleType(getReturnTypeFromContext(AFD));
```
Comment (rintaro) [inline lib/IDE/ExprContextAnalysis.cpp]: Please ensure this doesn't work for `init`/`deinit`. i.e.
```swift
enum MyEnum {
  case foo
  init() { .#^HERE^# }
}
```
should *not* work.

Comment (benlangmuir) [reply lib/IDE/ExprContextAnalysis.cpp]: Ah good point, for "init?" it should technically show `.none` and `nil`.

**Comment by benlangmuir** [comment]
@rintaro I added the tests for init, deinit, and top-level vars.  I was wrong about handling `.none` in a failable initializer; that doesn't work even with explicit `return`.  The only test case that isn't already working is 

```swift
init?() { #^A^# }
```

This case gets `nil`, but doesn't have the correct type context.  The reason is that the type-checker adds an implicit `return`  statement here, which doesn't happy for any other decls.  We could handle this if we cared, but it would make checking the body more complicated.  Since there is no legal unresolved members here it doesn't have any practical effect anyway, so I left it.

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please smoke test

**Pull request merged**